### PR TITLE
Roles stats & add ticks interval

### PR DESCRIPTION
**Pull request by Nevrdid**

### CODE REVIEW

**Comment by TooAngel** [comment]
DCBOT: 
This repository is under [democratic collaboration](https://github.com/TooAngel/democratic-collaboration) and will be merged automatically.

The merge decision is based on the outcome of the reviews:
 - `Approve` add the reviewer value (number of commits) to the `metric`
 - `Request changes` substract the reviewer value from the `metric`

 - `metric > 99%` merge now
 - `metric > 75%` merge in 1 day
 - `metric > 50%` merge in 3 days
 - `metric >= 0%` merge in 7 days

Please review the PR to make a good democratic decision.

This is the placeholder for reviewer summoning :-) (top two + random)

**Comment by TooAngel** [comment]
DCBOT: Current status percentage: 0 votes: 0 total: 229

**Comment by TooAngel** [comment]
DCBOT: Adding WIP label, the title is prefixed with [WIP]

**Comment by TooAngel** [comment]
DCBOT: Code update recognized, countdown starts fresh.

**Comment by TooAngel** [comment]
DCBOT: Removing WIP label, the title is not prefixed with [WIP]

**Comment by TooAngel** [comment]
DCBOT: Current status coefficient: 0.0 votes: 0 total: 234 age: 0

File: src/prototype_room_creepbuilder.js [hunk 1]
```diff
@@ -48,6 +48,10 @@ Room.prototype.spawnCreateCreep = function(role, heal, level, squad, routing, ba
       routing: routing
     };
     var returnCode = spawn.createCreep(partConfig, name, memory);
+    var userName = Memory.username || Game.rooms[Memory.myRooms[0]].controller.owner || 'default';
+    let roleStat = Memory.stats[Memory.username].roles[role];
+    let previousAmount = roleStat ? roleStat.amount : 0;
+    brain.stats.add('', '.roles.' + role, previousAmount + 1);
```
Comment (Somotaw) [inline src/prototype_room_creepbuilder.js]: src/config_brain_memory.js:44 has identical code? use that?

**Comment by TooAngel** [comment]
DCBOT: Current status coefficient: -0.0897435897436 votes: -21 total: 234 age: 0

**Comment by TooAngel** [comment]
DCBOT: Code update recognized, countdown starts fresh.

File: src/prototype_room_creepbuilder.js (hunk 1 above)
Comment (Somotaw) [inline src/prototype_room_creepbuilder.js]: `  let userName = Memory.username || Game.rooms[Memory.myRooms[0]].controller.owner || 'default';
    // Cleanup memory		    // Cleanup memory
    for (let name in Memory.creeps) {		    for (let name in Memory.creeps) {
      if (!Game.creeps[name]) {		      if (!Game.creeps[name]) {
       let role = Memory.creeps[name].role;
       let roleStat = Memory.stats[userName].roles[role];
       let previousAmount = roleStat ? roleStat.amount : 0;
       let amount = previousAmount ? 0 : previousAmount - 1;
       brain.stats.add('', '.roles.' + role, amount);
 `
is almost identical in brain_memory.jsL44

**Comment by TooAngel** [comment]
DCBOT: Current status coefficient: -0.0897435897436 votes: -21 total: 234 age: 0

**Comment by TooAngel** [comment]
DCBOT: Code update recognized, countdown starts fresh.

**Comment by TooAngel** [comment]
DCBOT: Current status coefficient: 0.0897435897436 votes: 21 total: 234 age: 0

**Comment by TooAngel** [comment]
DCBOT: Code update recognized, countdown starts fresh.

**Comment by TooAngel** [comment]
DCBOT: Current status coefficient: 0.0897435897436 votes: 21 total: 234 age: 0

**Comment by TooAngel** [comment]
DCBOT: Current status coefficient: 0.717948717949 votes: 168 total: 234 age: 0

**Comment by TooAngel** [comment]
DCBOT: Current status coefficient: 0.74358974359 votes: 174 total: 234 age: 0

File: src/config_brain_memory.js [hunk 2]
```diff
@@ -41,9 +41,17 @@ brain.prepareMemory = function() {
     console.log('Known constructionSites: ' + Object.keys(constructionSites).length);
   }
 
+  var userName = Memory.username || Game.rooms[Memory.myRooms[0]].controller.owner || 'default';
```
Comment (WhiteHalmos) [inline src/config_brain_memory.js]: let

File: src/config_brain_memory.js [hunk 3]
```diff
@@ -41,9 +41,17 @@ brain.prepareMemory = function() {
...
+  Memory.username = userName;
   // Cleanup memory
   for (let name in Memory.creeps) {
     if (!Game.creeps[name]) {
+      let role = Memory.creeps[name].role;
+      let roleStat = Memory.stats[userName].roles[role];
+      let previousAmount = roleStat ? roleStat.amount : 0;
+      let amount = previousAmount ? 0 : previousAmount - 1;
```
Comment (WhiteHalmos) [inline src/config_brain_memory.js]: +1 rather than -1?

Comment (WhiteHalmos) [inline src/config_brain_memory.js]: Opps sorry, missed this was cleanup.

**Comment by TooAngel** [comment]
DCBOT: Current status coefficient: 0.889361702128 votes: 209 total: 235 age: 1

**Comment by Somotaw** [comment]
Some issue still with the cleanup
4:31]TypeError: Cannot read property 'roles' of undefined
   at Object.brain.prepareMemory (config_brain_memory:50:44)
   at main (main:28:9)
   at Object.module.exports.loop (main:49:5)
   at __mainLoop:1:52

[23:26]  

[23:26:45]TypeError: Cannot read property '0

[23:26]  
TypeError: Cannot read property '0' of undefined
   at Object.brain.prepareMemory (config_brain_memory:44:62)
   at main (main:28:9)
   at Object.module.exports.loop (main:49:5)
   at __mainLoop:1:52

**Pull request merged**